### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by xiaozhong21**
## Overview

- Add dynamically rendered form to collection user portfolio information
- Update `My Stocks` page to render results of user's portfolios

## To-do
- Add `Edit` functionality to list of portfolios
- Implement more robust validation of `create portfolio` form (e.g. the constraint of asset allocations adding up to 100%)
- Historical stock charts for portfolios and individual stocks (on stock details page)

## Feedback Requested

Any feedback welcome!

Live on [Heroku](https://stocksimplified.herokuapp.com/)

### CODE REVIEW

**Comment by xiaozhong21** [comment]
@scabaral

File: server/src/portfolioRouter.mjs [hunk 1]
```diff
@@ -9,10 +9,39 @@ router.get("/", async (request, response) => {
...
+      request.params.portfolioID,
+      stock.ticker.toUpperCase(),
+      stock.allocation,
+    );
+  }
+  response.status(201).json(stocksArray);
+});
+
+router.delete("/:portfolioID", async (request, response) => {
+  await db.deletePortfolio(request.params.portfolioID);
```
Comment (scabaral) [inline server/src/portfolioRouter.mjs:43]: Nit: Use object destructuring

File: server/src/portfolioRouter.mjs [hunk 2]
```diff
@@ -9,10 +9,39 @@ router.get("/", async (request, response) => {
   response.json(portfolios);
 });
 
+router.get("/:portfolioID", async (request, response) => {
```
Comment (scabaral) [inline server/src/portfolioRouter.mjs:12]: These methods should probably have some error handling for if the requested object is not found or the write fails

File: server/src/db.mjs [hunk 3]
```diff
@@ -17,57 +40,41 @@ export const addOrUpdateUser = (user) =>
...
-
-export const addUserPortfolio = (sub) =>
-  db.one(
-    `INSERT INTO user_portfolio(user_id)
-      VALUES((SELECT id FROM users WHERE sub=$<sub>))
-      RETURNING *`,
-    { sub },
-  );
+export const deletePortfolio = (portfolioID) =>
+  db.none("DELETE FROM user_portfolio WHERE portfolio_id = $<portfolioID>", {
```
Comment (scabaral) [inline server/src/db.mjs:75]: When you delete a portfolio shouldn't you also delete the portfolio's stock? Or is postgres doing that for you with a `DELETE CASCADE`?

**Review by scabaral — COMMENTED** [review]
Great work Xiao! I love seeing all the progress you've made! I tried to leave helpful suggestions, let me know if you want to chat more about any of them! 

You're doing awesome! Keep it up!

File: app/src/AddPortfolio/index.js [hunk 4]
```diff
@@ -1,10 +1,221 @@
...
+    const oldVal = fields.length;
+    if (newVal > oldVal) {
+      for (let i = 0; i < oldVal; i++) {
+        update(i, {
+          ticker: watchFields[i],
+          allocation:
+            100 % newVal === 0 ? 100 / newVal : (100 / newVal).toFixed(2),
+        });
+      }
+      for (let i = oldVal; i < newVal; i++) {
```
[the remark this answers is not in the document]
Comment (xiaozhong21) [reply app/src/AddPortfolio/index.js:38]: Great question! 💯  The reason for having two for loops is that different operations will be needed for each loop: 0 to oldVal --> need to update existing pre-filled allocation numbers for each asset (using `update` method from react-hook-form); oldVal to newVal --> need to create new rows for additional asset inputs (using `append` method from react-hook-form).

File: server/src/portfolioRouter.mjs [hunk 5]
```diff
@@ -9,10 +9,39 @@ router.get("/", async (request, response) => {
...
 router.post("/", async (request, response) => {
-  const portfolio = await db.addUserPortfolio(request.user.sub);
+  const portfolio = await db.addUserPortfolio(request.user.sub, request.body);
   response.status(201).json(portfolio);
 });
 
+router.post("/:portfolioID/stocks", async (request, response) => {
+  const stocksArray = request.body;
+  for (let stock of stocksArray) {
+    await db.addPortfolioStocks(
```
[the remark this answers is not in the document]
Comment (xiaozhong21) [reply server/src/portfolioRouter.mjs]: Thank you for pointing this out, Sara! Not a straightforward fix using pg-promise but I learned a lot in the process of implementing bulk insert! https://github.com/xiaozhong21/StockSimplified/pull/8/commits/07815e9c8f26ee9ea7c84bb7bf8f86e7d227760f

File: app/src/MyStocks/Watchlist.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    </section>
+  ) : (
+    <section>
+      <h2>Watchlist</h2>
+      <button type="button" onClick={() => loadWatchlist()}>
+        Get real-time quotes
+      </button>
+      {watchlist.map((stock) => (
+        <li key={stock.symbol}>
+          <Link to={`/stocks/${stock.symbol}`}> {stock.symbol} </Link> |{" "}
```
[the remark this answers is not in the document]
Comment (xiaozhong21) [reply app/src/MyStocks/Watchlist.js:32]: This is just an Eslint auto correction thing. I will eventually remove it when I apply CSS on this page. 😄

File: server/src/db.mjs (hunk 3 above)
Comment (xiaozhong21) [reply server/src/db.mjs:85]: Yes! I have `ON UPDATE CASCADE ON DELETE CASCADE` in postgres.

**Pull request merged**